File: src/main/java/org/veupathdb/lib/blast/field/TBlastNTask.java
package org.veupathdb.lib.blast.field;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import com.fasterxml.jackson.databind.JsonNode;
import org.veupathdb.lib.blast.util.DefaultingJSONValue;
import org.veupathdb.lib.blast.util.JSONConstructor;

public enum TBlastNTask implements DefaultingJSONValue
{
  TBlastN("tblastn"),
  TBlastNFast("tblastn-fast");

  public static final TBlastNTask DefaultTask = TBlastN;

  private final String value;

  TBlastNTask(String value) {
    this.value = value;
  }

  public String getValue() {
    return value;
  }

  @Override
  public boolean isDefault() {
    return this == DefaultTask;
  }

  @Override
  @JsonValue
  public JsonNode toJSON() {
    return JSONConstructor.newText(value);
  }

  public static TBlastNTask fromString(String value) {
    for (var val : values())
      if (val.getValue().equals(value))
        return val;

    throw new IllegalArgumentException("Unrecognized task value \"" + value + "\".");
  }

  @JsonCreator
  public static TBlastNTask fromJSON(JsonNode js) {
    return fromString(js.textValue());
  }
}

File: src/main/java/org/veupathdb/lib/blast/field/TBlastNTask.kt
package org.veupathdb.lib.blast.field

import org.veupathdb.lib.blast.util.JSONConstructor.newText
import org.veupathdb.lib.blast.util.DefaultingJSONValue
import com.fasterxml.jackson.annotation.JsonValue
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.annotation.JsonCreator

enum class TBlastNTask(val value: String) : DefaultingJSONValue {
  TBlastN("tblastn"), TBlastNFast("tblastn-fast");

  override val isDefault: Boolean
    get() = this == DefaultTask

  @JsonValue
  override fun toJSON(): JsonNode {
    return newText(value)
  }

  companion object {
    @JvmStatic
    val DefaultTask = TBlastN

    @JvmStatic
    fun fromString(value: String): TBlastNTask {
      for (tmp in values())
        if (tmp.value == value)
          return tmp

      throw IllegalArgumentException("Unrecognized task value \"$value\".")
    }

    @JvmStatic
    @JsonCreator
    fun fromJSON(js: JsonNode): TBlastNTask {
      return fromString(js.textValue())
    }
  }
}